packages/svelte/src/compiler/phases/3-transform/client/visitors/Component.js

```diff
@@ -10,9 +10,10 @@ import { build_component } from './shared/component.js';
 export function Component(node, context) {
 	const component = build_component(
 		node,
-		// if it's not dynamic we will just use the node name, if it is dynamic we will use the node name
-		// only if it's a valid identifier, otherwise we will use a default name
-		!node.metadata.dynamic || regex_is_valid_identifier.test(node.name) ? node.name : '$$component',
+		// avoid shadowing the component variable by a variable used in $.component
+		node.metadata.dynamic
```

Comment:
Don't we need to keep the `regex_is_valid_identifier` test here for the `node.name` branch?

Also IMO we should cut this out and move it into its own variable -- it's getting hard to read

Reply:
The name should be kept as is for non-dynamic components.
In the case of dynamic components, the name just defines the name of an intermediate variable that was shadowing the component's one. And before #15118 it was always `$$component`. Here, the name only defines a more convenient stack tracing. The regex was there because the component name can be `Object.prop`, see `component-namespace` and `component-namespaced` tests.

I thought about separating the names of the component's source and the intermediate variable, but couldn't come up with a nice one. I can try again tomorrow.

Reply:
What if instead we just `context.scope.generate(node.name)`?